File: functions/schedule_manager/scheduler/update_schedule.py

from .store import get_path
from .store import set_path
from .schedule import Schedule
from .create_schedule import create_schedule
from .get_games import get_all_games


def update_schedule(key: str, year: int, post_week_buffer_hours: int):
    games = get_all_games(key, year)
    print(f"Fetched {year} data")

    previous_hash = read_schedule_hash(year)

    schedule = create_schedule(year, games, post_week_buffer_hours)

    if schedule.hash == previous_hash:
        print("No changes to schedule detected")
        return

    print("Schedule has changed, attempting to save...")

    save_schedule(year, schedule)

    print(f"{year} schedule has been updated")
    return schedule


def get_schedule_path(year: int) -> str:
    return f"schedule/{year}"


def read_schedule_hash(year: int) -> str:
    schedule_path = get_schedule_path(year)
    hash_path = f"{schedule_path}/hash"
    return get_path(hash_path)


def save_schedule(year: int, schedule: Schedule):
    path = get_schedule_path(year)
    set_path(path, schedule.serialize())

File: functions/schedule_manager/scheduler/update_schedule.go
package scheduler

import (
	"context"
	"fmt"
	"log"

	"github.com/mdryden/110yards/functions/schedulemanager/store"
)

func UpdateSchedule(key string, year int) error {
	log.Printf("Importing %d schedule", year)

	resp, err := GetGamesForSeason(key, year)
	if err != nil {
		return fmt.Errorf("failed to fetch games for %d: %v", year, err)
	}

	if resp == nil {
		return fmt.Errorf("failed to fetch games for %d: response was nil", year)
	}

	log.Printf("Fetched %d data", year)

	previousHash, err := ReadScheduleHash(year)
	if err != nil {
		return err
	}

	schedule, err := CreateSchedule(year, resp.Games)

	if err != nil {
		return err
	}

	if schedule.Hash == previousHash {
		log.Printf("No changes to schedule detected")
		return nil
	}

	log.Printf("Schedule has changed, attempting to save...")

	return SaveSchedule(year, schedule)
}

func getSchedulePath(year int) string {
	return fmt.Sprintf("schedule/%d", year)
}

func ReadScheduleHash(year int) (string, error) {
	schedulePath := getSchedulePath(year)
	hashPath := fmt.Sprintf("%s/hash", schedulePath)

	var hash string
	err := store.ReadPath(context.Background(), hashPath, &hash)

	if err != nil {
		return "", err
	}

	return hash, nil
}

func SaveSchedule(year int, schedule Schedule) error {
	path := getSchedulePath(year)
	return store.SetPath(context.Background(), path, schedule)
}
